### PR TITLE
close socket in terminate/2

### PR DESCRIPTION
**Pull request by id**
in some case when a user process got timeout and reconnected,
an old connection to postgres might stay hanging around for quite a while

### CODE REVIEW

**Comment by davidw** [comment]
I realize this is a tricky one to test, but the more tests, the better.

File: src/epgsql_sock.erl [hunk 1]
```diff
@@ -152,8 +152,10 @@ handle_info({inet_reply, _, Status}, State) ->
 handle_info({_, Sock, Data2}, #state{data = Data, sock = Sock} = State) ->
     loop(State#state{data = <<Data/binary, Data2/binary>>}).
 
+terminate(_Reason, #state{sock = Sock}) when is_port(Sock) ->
+    gen_tcp:close(Sock),
```
Comment (seriyps) [inline src/epgsql_sock.erl]: And what about ssl?

Comment (id) [inline src/epgsql_sock.erl]: Thanks, fixed.

File: src/epgsql_sock.erl [hunk 2]
```diff
@@ -152,9 +152,15 @@ handle_info({inet_reply, _, Status}, State) ->
 handle_info({_, Sock, Data2}, #state{data = Data, sock = Sock} = State) ->
     loop(State#state{data = <<Data/binary, Data2/binary>>}).
 
-terminate(_Reason, _State) ->
-    %% TODO send termination msg, close socket ??
-    ok.
+terminate(_Reason, #state{mod = Mod, sock = Sock}) ->
```
Comment (seriyps) [inline src/epgsql_sock.erl]: why not smth like

```
terminate(_, #state{sock=undefined}) -> ok;
terminate(_, #state{mod=gen_tcp}) -> gen_tcp:close(...);
terminate(..., ..mod=ssl}) -> ssl:close(...).
```
:smile:

**Comment by neeraj9** [comment]
Any specific reason this pull still open? According to me this is an important change.

**Comment by seriyps** [comment]
I'm ok with this.
@davidw ?

The only question is: AFAIK, when process, holding a connection dies, socket will be closed anyway. So, why implicitly close it?
Or, when process dies, socket closed in some other way, compared to explicit close?

**Comment by id** [comment]
Socket is not always closed immediately, it could stay open on OS level for a while even after your application restarted. Then the application will re-establish connection to postgres and it could lead to the situation when you exhausted limit on available connections for your user.

**Comment by seriyps** [comment]
Ok, let's wait @davidw for some time. For me it's ok.

**Comment by seriyps** [comment]
Thank you for patience and your work.

**Comment by davidw** [comment]
Sorry for the late reply but I wanted to say it looks like a good idea, and thanks!

**Comment by neeraj9** [comment]
Thanks. Saves me the trouble of keeping private merges.

**Pull request merged**